From dd6788253bf87662f62e831e2c7ac4f9d1784ccb Mon Sep 17 00:00:00 2001
From: Kaan Alsan <alsan@google.com>
Date: Tue, 20 Jan 2026 14:42:36 -0800
Subject: [PATCH] Paint dragged vertical tabs to a layer

This CL updates VerticalDraggedTabsContainer to explicitly paint dragged
views to a layer at the start of the drag (and destroying the layer at
the end) instead of relying on View::SetTransform to create the layer,
which had side effects with clipping.

Change-Id: I071b7b6b64a4ce3690d0a7758ab047dd1e86d389
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/7493706
Reviewed-by: Eshwar Stalin <estalin@chromium.org>
Commit-Queue: Kaan Alsan <alsan@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1571897}
---

--- a/chrome/browser/ui/views/tabs/vertical/vertical_dragged_tabs_container.cc
+++ b/chrome/browser/ui/views/tabs/vertical/vertical_dragged_tabs_container.cc
@@ -144,7 +144,6 @@ void VerticalDraggedTabsContainer::Initi
 void VerticalDraggedTabsContainer::ResetDragState() {
   for (auto view : dragging_views_) {
     view->SetTransform(gfx::Transform());
-    view->SetClipPath(view->clip_path());
   }
   UpdateLayoutForDrag();
   dragging_views_.clear();
@@ -160,9 +159,6 @@ void VerticalDraggedTabsContainer::Updat
     transform.Translate(
         0, GetYForDraggedTab(*tab_view, point_in_container, drag_clamp_min_y));
     tab_view->SetTransform(transform);
-    // Applying a transformation for the first time destroys the clip mask
-    // layer. Reapply the clip path in case.
-    tab_view->SetClipPath(tab_view->clip_path());
   }
 }
 
--- a/chrome/browser/ui/views/tabs/vertical/vertical_tab_drag_handler.cc
+++ b/chrome/browser/ui/views/tabs/vertical/vertical_tab_drag_handler.cc
@@ -23,6 +23,7 @@
 #include "components/tabs/public/tab_group_tab_collection.h"
 #include "components/tabs/public/tab_interface.h"
 #include "ui/base/metadata/metadata_impl_macros.h"
+#include "ui/compositor/layer.h"
 #include "ui/views/view_utils.h"
 
 namespace {
@@ -339,6 +340,8 @@ void VerticalTabDragHandlerImpl::Started
     auto* shim_view = views::AsViewClass<TabSlotShimView>(view);
     CHECK(shim_view);
     dragged_tabs_.insert(&shim_view->node());
+    shim_view->parent()->SetPaintToLayer();
+    shim_view->parent()->layer()->SetFillsBoundsOpaquely(false);
   }
 }
 
@@ -347,6 +350,9 @@ void VerticalTabDragHandlerImpl::Dragged
 }
 
 void VerticalTabDragHandlerImpl::StoppedDragging() {
+  for (auto& [_, shim_view] : shim_views_) {
+    shim_view->parent()->DestroyLayer();
+  }
   dragged_tabs_.clear();
 }
 
